Added `linkerd uninject` command

alpeb opened the pull request:
uninject.go iterates through the resources annotations, labels,
initContainers and Containers, removing what we know was injected by
linkerd.

The biggest part of this commit is the refactoring of inject.go, to make
it more generic and reusable by uninject. Some common parts were moved
into inject_util.go but others common parts remained in inject.go to
ease review and will then later get moved (marked with a comment).

The entry-point function `transformInput` (formerly called
`runInjectCmd`) receives as a parameter the function holding the logic
specific for injection (`injectResource`) or uninjection
(`uninjectResource`).

`injectResource` was split into the `parse` (reused by uninject) and
`injectResource` methods on the new `resourceConfig` receiver. That
receiver holds all the info required for the injection/uninjection.

The idea is that in a following PR this functionality will get reused by
`linkerd inject` to uninject as as preliminary step to injection, as a
solution to #1970.

This was tested successfully on emojivoto with:

```
1) inject:
kubectl get -n emojivoto deployment -o yaml | bin/linkerd inject - |
kubectl apply -f -
2) uninject:
kubectl get -n emojivoto deployment -o yaml | bin/linkerd uninject - |
kubectl apply -f -
```

wmorgan commented:
I assume this should be "add `linkerd uninject` command" :)

grampelberg commented:
```
curl https://run.linkerd.io/emojivoto.yml | bin/go-run cli inject --tls=optional -  | bin/go-run cli uninject
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3238  100  3238    0     0   6425      0 --:--:-- --:--:-- --:--:--  6424
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/linkerd/linkerd2/cli/cmd.newCmdUninject.func1(0xc00002d200, 0x3711820, 0x0, 0x0, 0x0, 0x0)
	/Users/thomas/go/src/github.com/linkerd/linkerd2/cli/cmd/uninject.go:27 +0x1a2
github.com/linkerd/linkerd2/vendor/github.com/spf13/cobra.(*Command).execute(0xc00002d200, 0x3711820, 0x0, 0x0, 0xc00002d200, 0x3711820)
	/Users/thomas/go/src/github.com/linkerd/linkerd2/vendor/github.com/spf13/cobra/command.go:698 +0x4fa
github.com/linkerd/linkerd2/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x36e57a0, 0xc00002c000, 0xc000437f70, 0x2489d90)
	/Users/thomas/go/src/github.com/linkerd/linkerd2/vendor/github.com/spf13/cobra/command.go:783 +0x3bd
github.com/linkerd/linkerd2/vendor/github.com/spf13/cobra.(*Command).Execute(0x36e57a0, 0x2577d20, 0x2810c90)
	/Users/thomas/go/src/github.com/linkerd/linkerd2/vendor/github.com/spf13/cobra/command.go:736 +0x39
main.main()
	/Users/thomas/go/src/github.com/linkerd/linkerd2/cli/main.go:10 +0x4b
```

alpeb commented:
@grampelberg Nice catch. I had missed validating the presence of the config-file argument (no dash after `bin/go-run cli uninject`)

grampelberg commented:
I don't think `check` should run on uninject. Otherwise, it is working great for me!

alpeb commented:
@grampelberg if you mean like the preliminary checks that `linkerd stat` or `linkerd dashboard` perform, it's not. The validation I mentioned above is just a simple sanity verification to ensure something is passed onto `linkerd uninject`, just like we do for `linkerd inject`.
Thanks for the tests! :slightly_smiling_face:

grampelberg commented:
I get this output when I run `uninject`:

```
        name: linkerd-trust-anchors
      - name: linkerd-secrets
        secret:
          optional: true
          secretName: vote-bot-deployment-tls-linkerd-io
status: {}
---

hostNetwork: pods do not use host networking...............................[ok]
sidecar: pods do not have a proxy or initContainer already injected........[ok]
supported: at least one resource injected..................................[ok]
udp: pod specs do not include UDP ports....................................[ok]

Summary: 4 of 8 YAML document(s) injected
  deployment/emoji
  deployment/voting
  deployment/web
  deployment/vote-bot
```

I would expect the `hostNework` checks to not be there and the summary to say `4 of 8 document(s) uninjected`.

grampelberg commented:
Also, I just noticed that uninject didn't get rid of everything (the volumes)!

alpeb commented:
@grampelberg I had forgotten about that report at the end; you're right, I'll get rid of that. Will also address the volumes removal :+1:

alpeb commented on cli/cmd/inject.go:
```diff
@@ -46,12 +45,15 @@ type injectOptions struct {
+// InjectYAML processes resource definitions and outputs them after injection in out
+var InjectYAML = func(in io.Reader, out io.Writer, report io.Writer, options *injectOptions) error {
+	return ProcessYAML(in, out, report, options, resourceTransformerInject{})
+}
```
I have no idea why I did that :man_shrugging:  Will use a regular func.

alpeb commented on cli/cmd/uninject.go:
```diff
@@ -0,0 +1,152 @@
+You can use a config file from stdin by using the '-' argument
+with 'linkerd uninject'. e.g. curl http://url.to/yml | linkerd uninject -
+Also works with a folder containing resource files and other
+sub-folder. e.g. linkerd uninject <folder> | kubectl apply -f -
```
Agreed. This was based on the `Long` field from `linkerd inject`. I'll make the same change there for consistency sake.

alpeb commented on cli/cmd/uninject.go:
```diff
@@ -0,0 +1,152 @@
+	t.Annotations = newAnnotations
+
+	labels := make(map[string]string)
+	ldLabels := []string{k8s.ControllerNSLabel}
```
In the following lines it expands that slice with the labels that are specific to the current resource kind being processed. Or were you implying that we should simply iterate through all the labels regardless of the resource kind?

klingerf replied:
Ah, ok, that makes sense. I was testing this by running inject on a deployment, and then running uninject on a single pod from the deployment. For instance:

```
bin/linkerd install | kubectl apply -f -
bin/linkerd inject test/testdata/smoke_test.yaml | kubectl apply -f -
kubectl get po/smoke-test-gateway-85db44995d-8krxg -oyaml | bin/linkerd uninject -
```

And in the uninjected YAML, I still see:

```
  labels:
    app: smoke-test-gateway
    linkerd.io/proxy-deployment: smoke-test-gateway
    pod-template-hash: "4186005518"
```

Which makes sense, since the inject resource type was "deployment", but the uninject resource type was "pod". I can't think of a better fix for this than iterating through all of the labels, but I agree that it feels heavy handed.

alpeb replied:
I hadn't realized how those labels propagate to pods, so the wholesale approach makes sense. I'll apply that :+1:

grampelberg commented:
Just tested with some super complex resources and everything worked! We're doing some weird altering of YAML but I don't think that's a huge deal in the long run (and doesn't appear to be related to this issue).

alpeb commented:
Thanks @grampelberg for the thorough testing!

@klingerf These are the changes in the latest push:

- Modified inject's and uninject's `Long` and `Example` fields.
- Created the `InjectedLabels` slice in labels.go, that we use in uninject.go to make sure all the labels are removed.
- Reused the `sidecar` property during uninjection, to properly track the number of resources that were actually uninjected, for the report to make more sense.
- Created unit tests for uninject.go. The fixture files from the inject tests could be reused. But as now the input files act as outputs, they represent existing resources and required these changes (that didn't affect inject):
  - Rearranged fields in alphabetical order.
  - Added fields that are only relevant for existing resources (e.g. creationTimestamp and status.replicas in StatefulSets)
    
Still pending: move common functionality into inject_util.go
